### PR TITLE
Hybrid deployment in Tyk Cloud

### PR DESCRIPTION
**Pull request by SonjaChevre**
## Related Issue (if applicable)

n/a

## Description

Improved the page https://tyk.io/docs/tyk-cloud/environments-deployments/hybrid-gateways/ to show how to deploy Hybrid Gateways and connect them to Tyk Cloud. 

## Motivation and Context

First step in helping our users deploy Hybrid in Tyk Cloud. More improvements on the product side coming soon. 

## Preview Link

https://deploy-preview-2397--tyk-docs.netlify.app/docs/nightly/tyk-cloud/environments-deployments/hybrid-gateways/

## Types of changes

- [ ] Fixing typo (please merge to production)
- [ ] Documenting a new feature (please merge to production)
- [ ] Documentation for future release (please do not merge to production)
- [x] Something else: improving documentation for an existing feature, should be merged to production

## Checklist

- [x] I have [reviewed the guidelines](../CONTRIBUTING.md) for contributing to this repository.
- [x] I have [read the technical guidelines](../CONTRIBUTING-TECHNICAL-GUIDE.md) for contributing to this repository.
- [x] Make sure you have started *your change* off *our latest `master`*.
- [x] I have added a preview link.

### CODE REVIEW

**Comment by letzya** [comment]
changed for you to "draft" as in WIP

File: tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md [hunk 1]
```diff
@@ -1,78 +1,205 @@
...
-1. Add the Tyk official Helm repo `tyk-helm` to your local Helm registry
+**1. Add the Tyk official Helm repo `tyk-helm` to your local Helm registry**
 
 ```bash
 helm repo add tyk-helm https://helm.tyk.io/public/helm/charts/
 helm repo update
 ```
 
-2. Before proceeding with installation of the chart we need to set some custom values. First save the full original values.yaml to a local copy:
+The helm charts are also available on [ArtifactHub](https://artifacthub.io/packages/helm/tyk-helm/tyk-pro).
```
Comment (carolwkc) [inline tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md]:
```suggestion
The helm charts are also available on [ArtifactHub](https://artifacthub.io/packages/helm/tyk-helm/tyk-hybrid).
```

File: tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md [hunk 2]
```diff
@@ -1,78 +1,205 @@
...
-- `gateway.rpc.apiKey`, 
-- `gateway.rpc.connString`.
+**4. Configure Tyk Gateway and its connection to Tyk Cloud**
+
+You need to modify the following values in your custom `values.yaml` file: 
+
+* `gateway.rpc.apiKey`: Tyk Dashboard API Access Credentials of the user created ealier
+* `gateway.rpc.rpcKey`: Organisation ID 
+* `gateway.rpc.connString`: MDCB connection string
+* *(optional)* `gateway.rpc.groupId`: if you have multiple data plane (e.g. in different regions), specify the data plane group (string) to which the gateway you are deploying belong
```
Comment (carolwkc) [inline tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md]:
```suggestion
* *(optional)* `gateway.rpc.groupId`: if you have multiple data plane (e.g. in different regions), specify the data plane group (string) to which the gateway you are deploying belong. The data planes in the same group share one redis.
```

File: tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md [hunk 3]
```diff
@@ -1,78 +1,205 @@
...
+* *(optional)* `gateway.sharding.enabled` and `gateway.sharding.tags`: you can enable sharding to selectively load APIs to specific gateways, using tags
+* `image.tag`: `v4.3` (or any other version available in [dockerhub](https://hub.docker.com/r/tykio/tyk-gateway/tags))
```
Comment (carolwkc) [inline tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md]:
```suggestion
* `gateway.image.tag`: `v4.3` (or any other version available in [dockerhub](https://hub.docker.com/r/tykio/tyk-gateway/tags))
```

Comment (carolwkc) [reply tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md]: I think the page https://hub.docker.com/r/tykio/tyk-gateway/tags is a bit hard to follow. The tags are not ordered/filtered and there's no release information. Shall we consider adding image tag information to https://tyk.io/docs/release-notes/version-4.3/ @letzya ?

Got a look at https://hub.docker.com/_/kong and it looks more friendly. There's a list of currently supported image tags on Overview page. We can list or LTS and latest version there. I'll raise a separate suggestion to DevOps team :)

Comment (letzya) [reply tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md]: Yes, we should definitely update our release notes.

File: tyk-docs/assets/img/hybrid-gateway/tyk-cloud-dashboard-api-manager.png
Comment (carolwkc) [inline tyk-docs/assets/img/hybrid-gateway/tyk-cloud-dashboard-api-manager.png:1]: shall we mask the name & url on this page too?

File: tyk-docs/assets/img/hybrid-gateway/tyk-cloud-hybrid-connection-control-plane.png
Comment (carolwkc) [inline tyk-docs/assets/img/hybrid-gateway/tyk-cloud-hybrid-connection-control-plane.png:1]: shall we mask the name & url on this page too?

File: tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md [hunk 4]
```diff
@@ -1,158 +1,299 @@
...
+* Tyk Cloud Account, register here if you don't have one yet: {{< button_left href="https://tyk.io/sign-up/#cloud" color="green" content="free trial" >}}
+* A redis instance for each data plane, used as temporay storage for distributed rate limiting, token storage and analytics. You will find instructions for a simple Redis installation in the steps below. 
+* No incoming firewalls rules are needed, as the connection between Hybrid Gateways and Tyk Cloud is always initiated from the Gateways, not from Tyk Cloud.
 
+## Get the connection details to the control plane
 
-## Hybrid Gateways in a Kubernetes Cluster
-This Helm Chart provides a method of adding Hybrid Gateways into your Kubernetes cluster.
-The Hybrid Gateways can be connected to *Tyk Cloud* or to a *Tyk Self managed Control plane* (a.k.a *MDCB*/*Tyk Multi Data Centre Bridge (MDCB)*).
+Before deploying your data plane, you will need to prepare the connection details to the control plane:
```
Comment (carolwkc) [inline tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md]:
```suggestion
The hybrid data plane can connect to control plane in Tyk Cloud by using the Tyk Dashboard API Access Credentials. Follow the guides below to create a user (identity for hybrid data plane) and prepare the connection details:
```

suggest to give some context about why users need to create a user below

File: tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md [hunk 5]
```diff
@@ -1,158 +1,299 @@
...
+  {{< img src="/img/hybrid-gateway/tyk-cloud-dashboard-api-manager-user.png" alt="API Manager Dashboard" >}}
 
-If you are interested in contributing to our charts, suggesting changes, creating PRs or any other way,
-please use [GitHub Tyk-helm-chart repo](https://github.com/TykTechnologies/tyk-helm-chart/tree/master/tyk-hybrid)
+  - Copy the **Tyk Dashboard API Access Credentials** for later use (`gateway.rpc.apiKey` setting in helm)
```
Comment (carolwkc) [inline tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md]:
```suggestion
  - Copy the **Tyk Dashboard API Access Credentials** for later use (`api_key` setting in Docker, `gateway.rpc.apiKey` setting in helm)
```

File: tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md [hunk 6]
```diff
@@ -1,158 +1,299 @@
...
+  - Copy the **Organisation ID** for later use (`gateway.rpc.rpcKey` setting in helm)
```
Comment (carolwkc) [inline tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md]:
```suggestion
  - Copy the **Organisation ID** for later use (`rpc_key` setting in Docker, `gateway.rpc.rpcKey` setting in helm)
```

File: tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md [hunk 7]
```diff
@@ -1,158 +1,299 @@
...
 ```
 
-3. For tyk-hybrid chart we need to modify the following values in your custom `values.yaml` file: 
-- `gateway.rpc.rpcKey`, 
-- `gateway.rpc.apiKey`, 
-- `gateway.rpc.connString`.
+## Deploy in Kubernetes with Helm
 
-You can get the values from API Manager Dashboard.
+### 1. Add the Tyk official Helm repo `tyk-helm` to your local Helm registry
```
Comment (carolwkc) [inline tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md]:
```suggestion
### 1. Add the Tyk official Helm repo `tyk-helm` to your local Helm repository
```

File: tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md [hunk 8]
```diff
@@ -1,158 +1,299 @@
...
-4. Then we can install the chart using our custom values file:
+### 2. Then create a new namespace that will be hosting the Tyk Gateways
 
 ```bash
-helm install tyk-hybrid tyk-helm/tyk-hybrid -f values.yaml -n tyk
+kubectl create namespace tyk
 ```
 
-Check this (doc)[/tyk-multi-data-centre/setup-slave-data-centres/] for detailed explanation of the hybrid/worker Gateway settings.
+### 3. Get the template yaml for configuration
```
Comment (carolwkc) [inline tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md]:
```suggestion
### 3. Get the default values.yaml for configuration
```

File: tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md [hunk 9]
```diff
@@ -1,158 +1,299 @@
...
+### 4. Configure Tyk Gateway and its connection to Tyk Cloud
 
-Although these instructions are for our containerized Gateway, the required configuration changes are the same regardless of how you’re running your Gateways (Bare metal, VM, etc.), you should update the <tyk.conf> for your Gateway install instead of <tyk.hybrid.conf>
+You need to modify the following values in your custom `values.yaml` file: 
 
-{{< /note >}}
+* `gateway.rpc.apiKey`: Tyk Dashboard API Access Credentials of the user created ealier
+* `gateway.rpc.rpcKey`: Organisation ID 
+* `gateway.rpc.connString`: MDCB connection string
+* *(optional)* `group_id`: if you have multiple data plane (e.g. in different regions), specify the data plane group (string) to which the gateway you are deploying belong. The data planes in the same group share one redis instance.
```
Comment (carolwkc) [inline tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md]:
```suggestion
* *(optional)* `gateway.rpc.group_id`: if you have multiple data plane (e.g. in different regions), specify the data plane group (string) to which the gateway you are deploying belong. The data planes in the same group share one redis instance.
```

File: tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md [hunk 10]
```diff
@@ -1,158 +1,299 @@
...
+* *(optional)* `gateway.sharding.enabled` and `gateway.sharding.tags`: you can enable sharding to selectively load APIs to specific gateways, using tags
```
Comment (carolwkc) [inline tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md]:
```suggestion
* *(optional)* `gateway.sharding.enabled` and `gateway.sharding.tags`: you can enable sharding to selectively load APIs to specific gateways, using tags. By default, sharing is disabled and gateway will load all APIs.
```

File: tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md [hunk 11]
```diff
@@ -1,158 +1,299 @@
...
-### Steps for installation
+  export REDIS_PASSWORD=$(kubectl get secret --namespace tyk tyk-redis -o jsonpath="{.data.redis-password}" | base64 --decode)
+```
 
-1. Firstly, clone all repo files.
+You need to modify the following values in your custom `values.yaml` file: 
 
-{{< img src="/img/hybrid-gateway/image3-35.png" alt="Clone repo files" >}}
+* `redis.addrs`: the name of the redis instance including the port as set by Bitnami `tyk-redis-master.tyk.svc.cluster.local:6379` 
+* `redis.pass`: password set in redis (`$REDIS_PASSWORD`)
```
Comment (carolwkc) [inline tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md]:
```suggestion
* `redis.pass`: password set in redis (`$REDIS_PASSWORD`). Alternatively, you can use --set flag to set it during helm installation. For example --set redis.pass=$REDIS_PASSWORD.
```

File: tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md [hunk 12]
```diff
@@ -1,158 +1,299 @@
...
 
-2. Follow the docs in the repo, there's a [tyk.hybrid.conf](https://github.com/TykTechnologies/tyk-gateway-docker#hybrid) file that needs to be configured with the appropriate configuration items. To change these, head to your Tyk Cloud account. You need to change the following three values in **<tyk.hybrid.conf>**
 
-```json
-"slave_options": {
-"rpc_key": "<ORG_ID>",
-"api_key": "<API-KEY>",
-"connection_string": "<MDCB-INGRESS>:443",
-``` 
+### 6. Install the helm chart
```
Comment (carolwkc) [inline tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md]:
```suggestion
### 6. Install Hybrid data plane
```

File: tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md [hunk 13]
```diff
@@ -1,158 +1,299 @@
...
+Note: IP adresses might differ on your system. 
 
-```bash
-docker compose up -d
+
+Finally, from your console, send a HTTP call to the /hello endpoint of the gateway `gateway-svc-tyk-hybrid`:
+
+```console
+curl http://hostname:32668/hello -i
+````
```
Comment (carolwkc) [inline tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md]:
```suggestion
Note: you may need to port forward if you're testing on local machine, e.g. `kubectl port-forward service/gateway-svc-tyk-hybrid -n tyk 8080:443`

```console
curl http://hostname:8080/hello -i
````
```

File: tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md [hunk 14]
```diff
@@ -1,158 +1,299 @@
...
+    "redis": {"status":"pass","componentType":"datastore","time":"2023-03-15T11:39:10Z"},
+    "rpc": {"status":"pass","componentType":"system","time":"2023-03-15T11:39:10Z"}}
+}
 ```
 
-You should now have two running containers, a Gateway and a Redis.
+
+## Next steps
 
 Now it is time to publish a new API [Task 5 - Deploy your Edge Gateway and add your first API]({{< ref "/content/tyk-cloud/getting-started-tyk-cloud/first-api.md" >}})
```
Comment (carolwkc) [inline tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md:300]: is this link correct?

**Review by carolwkc — CHANGES_REQUESTED** [review]
Thanks for the updates

File: tyk-docs/assets/img/hybrid-gateway/tyk-cloud-dashboard-api-manager.png
Comment (SonjaChevre) [reply tyk-docs/assets/img/hybrid-gateway/tyk-cloud-dashboard-api-manager.png:1]: good call - fixed

File: tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md (hunk 14 above)
Comment (SonjaChevre) [reply tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md:300]: yes, it goes to the page: https://tyk.io/docs/tyk-cloud/getting-started-tyk-cloud/first-api/ 
The page needs also to be improved in the future (e.g. removing "edge gateway" and making it more about configuring APIs in Tyk Cloud using Tyk Dashboard).

**Review by carolwkc — APPROVED** [review]
LGTM

**Comment by SonjaChevre** [comment]
@letzya can you please merge?

File: tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md [hunk 15]
```diff
@@ -1,158 +1,300 @@
...
 ```
 
-In this compose file, we've now got our gateway image, we've got Redis and we have some volume mappings.
-Run the followng:
+Note: IP adresses might differ on your system. 
 
-```bash
-docker compose up -d
+
+Finally, from your console, send a HTTP call to the /hello endpoint of the gateway `gateway-svc-tyk-hybrid`:
```
Comment (letzya) [inline tyk-docs/content/tyk-cloud/environments-deployments/hybrid-gateways.md:272]:
```suggestion
Finally, from your terminal, send an HTTP call to the `/hello` endpoint of the gateway `gateway-svc-tyk-hybrid`:
```
